updated information and link extractor

shaibrown5 opened the pull request.

shaibrown5 committed:
This script executes 3 tasks:
1. webscrape the url's from the given url list and find the zip file
link
2. extract the xml file from the zip and parse it
3. extract the relevant info from the xml tree and place into the
relevant database tables

very slow as it has to go through almost all the links,
sql table input is an example only used for 2 links

shaibrown5 committed:
Update information_extractor.py

shaibrown5 committed:
Merge branch 'master' of https://github.com/beyond-io/superx

shaibrown5 committed:
removed item price from info table

shaibrown5 committed:
changes zip_fils into a set, added .decode method

after some debugging saw that multiples if not more of the saem link were entered into the zip_links , and so I changed zip_links into a set

shaibrown5 committed:
added a third super market and made relevant changes to accomidate it

shaibrown5 committed:
Merge branch 'master' of https://github.com/shaibrown5/superx

shaibrown5 committed:
Merge branch 'master' of https://github.com/beyond-io/superx

shaibrown5 committed:
dded a third supermarket and made relevant changes

shaibrown5 committed:
minor changes to the third super

shaibrown5 committed:
minor changes

simzacks commented on superx/information_extractor.py:
```diff
@@ -0,0 +1,131 @@
+
+# url list with category PriceFull
+url_list = ['http://publishprice.mega.co.il/%s' % datetime.today().strftime('%Y%m%d'),  # mega link with correct date
+            'http://prices.shufersal.co.il/FileObject/UpdateCategory?catID=2&storeId=0&sort=Category&sortdir=ASC&page=1',
```
All the shufersal links are the same aside from the page. Instead of listing each page individually here, other ways to do it would be to have an additional variable that says how many pages, or see what happens when you get to a non-existent page and look for that error.
You can also have a dictionary/variable for each chain defining if it is a single page or multiple pages that need to be scrolled through

shaibrown5 replied:
there are 23 pages, if for instance i ask for page 55 then i get page 23, so there is no harm done asking for more pages than given. Ive checked on multiple days and all have had 23 pages. I could make an external method that all it does is add 3 shufersal url's to the list, but i feel that this is redundant. The urls are constants and so why not build a list that is created with its variables?

shaibrown5 replied:
with regards to the rest, they use only 1 page. (the rest meaning the other links that i have implemented into the code

simzacks replied:
I'm not going to hold up the PR for this, but it's not good looking code. In this case, most probably the number of pages won't change, but they might at some point in time and you wouldn't know.

shaibrown5 replied:
ok i understand and ill think of a way to make it more general

simzacks commented on superx/information_extractor.py:
```diff
@@ -0,0 +1,131 @@
+        if super_name == 'mega':
+            zip_link = url_list[0] + '/' + zip_link
+        elif super_name == 'victory':
+            zip_link = 'http://matrixcatalog.co.il/' + zip_link
```
in your url list, victory is listed as: http://matrixcatalog.co.il/NBCompetitionRegulations.aspx
This code will add "http://matrixcatalog.co.il/" to the front of that.

shaibrown5 replied:
i think you misunderstood this, the zip_link is not the link of the url, it is the link of the zip file online. the pages HTML does only has the suffix of the link so it is necessary to add the proper prefix

simzacks replied:
definitely misunderstood

simzacks commented on superx/information_extractor.py:
```diff
@@ -0,0 +1,131 @@
+            item_id = 0
+
+        item_info_formula = """INSERT INTO item_info (itemCode,itemId,itemName) VALUES(%s,%s,%s) ON DUPLICATE KEY UPDATE itemName = itemName """
+        price_table_shufersal = """INSERT INTO price_table (itemCode, shufersal) VALUES(%s, %s)  ON DUPLICATE KEY UPDATE shufersal = %s """
```
This looks like your price table has a column for every supermarket chain, where you are putting the price.
The price can be different at every store as well. So this design will not work well.
Also you want to avoid writing seperate code for each supermarket, if you can avoid it.

simzacks commented on superx/information_extractor.py:
```diff
@@ -0,0 +1,131 @@
+
+        # places the info into the the item_info table
+        cursor.execute(item_info_formula, (code, item_id, name))
+        conn.commit()
```
Why commit here if you are committing below? You probably want both statements to go through together.

shaibrown5 replied:
ok i thought .commit() was per statement

simzacks replied:
commit is when you want to save to the database. So all statements that you want either them all to go in or none, should be in one commit.

simzacks commented on superx/information_extractor.py:
```diff
@@ -0,0 +1,131 @@
+        except AttributeError:
+            item_id = 0
+
+        item_info_formula = """INSERT INTO item_info (itemCode,itemId,itemName) VALUES(%s,%s,%s) ON DUPLICATE KEY UPDATE itemName = itemName """
```
Instead of having a lot of insert statements, you might want to consider building all the valkues into a single insert statement.
values = ""
for item in items.findall(item):
    ...
    values_item = values_item +"(%s,%s,%s)," % ...
    values_price = values_price +"(%s,%s,%s)," % ...
    ...
values_item = values_item[:-1] # to remove the last comma
values_price = values_price[:-1] # to remove the last comma

and then do the cursor.execute once per statement. 
There might be a character limit on how much you can send to the server at a time.

Another option is not to commit until you have done all the inserts, so you don't have as many transactions
There might be

shaibrown5 replied:
this is no longer relevant as we have moved to sql alchemy and so the queries will be different

shaibrown5 committed:
Merge branch 'master' of https://github.com/beyond-io/superx

shaibrown5 committed:
Merge branch 'master' of https://github.com/beyond-io/superx

shaibrown5 committed:
Merge remote-tracking branch 'origin'

shaibrown5 committed:
wrapped the script ina class as i think this is best.
added a few new functions and updated which data i get from the xml file
so that it matches the db tables.

shaibrown5 commented on superx/information_extractor.py:
```diff
@@ -9,123 +9,155 @@
+            except AttributeError:
+                item_id = 0
+
+            item_info_formula = """INSERT INTO item_info (itemCode,itemId,itemName) VALUES(%s,%s,%s) ON DUPLICATE KEY UPDATE itemName = itemName """
```
this is no longer relevant as we moved to sql alchemy, i have yet to update the queries

shaibrown5 committed:
This is the final version of the information extractor script
updated the database to sqlalchemy
added a couple of helper function
added comments

shaibrown5 committed:
updated import line

simzacks commented on superx/information_extractor.py:
```diff
@@ -0,0 +1,201 @@
+                         ]
+        # current supermarket name
+        self.super_name = ''
+        db.create_all()
```
I would move db creation to either models/__init__.py or app.py 
This part of the app should be able to assume that the database exists

simzacks commented on superx/information_extractor.py:
```diff
@@ -0,0 +1,201 @@
+        elif 'shufersal' in url:
+            self.super_name = 'shufersal'
+        elif 'matrix' in url:
+            self.super_name = 'victory'
```
Just a suggestion:
to make this more concise (and make it easier to add new supermarkets),
you can do either:
( if there is no need to change the name from matrix to victory)
```
supers = ['mega', 'shufersal', 'matrix']
for super in supers:
    if super in url:
        self.super_name = super
        break
or 
supers = {"mega": "mega", "shufersal": "shufersal", "matrix": "victory"}
for super in supers:
    if super in url:
        self.super_name = supers[super]
        break
```

simzacks replied:
If you have multiple if statements in your code for variances in each store, you can make a dictionary of dictionaries, and define it all at the beginning. This will also make adding new supermarkets easier.
As an example:
```
supers = {"mega": {"store_name": "mega", "zip_link_prefix": "http://..." },
"shufersal": {"store_name": "shufersal"},
"matrix": {"store_name: "victory", "zip_link_prefix": "http://matrixcatalog.co.il/", "item": "Product", "bIsWeighted": "BisWeighted", "date_format": "%Y/%m/%d"}}
```
You could then use code such: 
`supers[self.super_name].get("item", "Item") #default to Item if it isn't specially defined`
and
`supers[self.super_name].get("zip_link_prefix", "")`

shaibrown5 replied:
I have added a dictionary of dictionaries, in the new version and will commit it shortly

simzacks commented on superx/information_extractor.py:
```diff
@@ -0,0 +1,201 @@
+
+
+class info_extractor:
+    """
```
python naming convention suggests that class names should use CapWords convention. The class name would be InfoExtractor
https://www.python.org/dev/peps/pep-0008/#class-names

simzacks commented on superx/information_extractor.py:
```diff
@@ -0,0 +1,201 @@
+
+        # add all the product info to the db and commit
+        db.session.add_all(products_list)
+        db.session.commit()
```
you probably want to use only the commit below as you want all this data to go in together

shaibrown5 committed:
changed nameing convention and othe rminor changes

shaibrown5 committed:
changed the way we commit, instead of a list, each individual product

added names of product to be excluded
minor changes

shaibrown5 committed:
Complete change of the script:
removed url_list,
added a multi level dictionary that hold the needed attributes for the script. this allows for ease when adding a new supermarket.
The script now runs through the supermarkets in the dictionary rather than a list of urls

furthermore:
- added a method for getting the amount of pages - currently only works for shufersal
- added a method for creating a url_list based on the ammount of pages needed
- exception handling: incase request fails, the current loop moves to the next iteration rather than stopping the whole script

shaibrown5 committed:
fixed an if statement

shaibrown5 committed:
fixed a variable name

shaibrown5 committed:
Merge branch 'master' of github.com:beyond-io/superx

simzacks commented on superx/information_extractor.py:
```diff
@@ -22,133 +22,139 @@ class info_extractor:
+            url_list = [self.current_super['url']]
+
+            if self.current_super['multiple_pages']:
+                url_list = self.get_all_super_links()
```
1000 times cleaner. ++++++1

simzacks commented on superx/information_extractor.py:
```diff
@@ -22,133 +22,139 @@ class info_extractor:
+        self.current_super = ''
+        # list of unwanted names to be filter out
+        self.exclude_names = ['משלוחים', 'ריק', 'פיקדון', 'תיבה', 'משלוח']
+        self.super_specific_information = {'mega': {'store_name': 'mega',
```
this is very nice.

simzacks replied:
In the future you might even consider putting this into the database, and then to add a new supermarket, you don't need to touch the code, just add a new row. But lets not think about that yet.

shaibrown5 replied:
I will talk with oded about it

simzacks commented on superx/information_extractor.py:
```diff
@@ -22,133 +22,139 @@ class info_extractor:
+                page = requests.get(url)
+                web_scrapper = BeautifulSoup(page.content, 'html.parser')
+            except ConnectionError:
+                continue
```
If there's a connection error, it will just ignore it and you won't know that you had a problem. Have you considered logging actions/problems?

simzacks commented on superx/information_extractor.py:
```diff
@@ -22,133 +22,139 @@ class info_extractor:
+                request = requests.get(zip_link)
+                content = request.content
+            except ConnectionError:
+                continue
```
have you considered logging errors so you will know when there is a problem?

simzacks commented on superx/information_extractor.py:
```diff
@@ -22,133 +22,139 @@ class info_extractor:
+                items = tree.getchildren()[-1]
+                self.extract_information(items)
+
+    def extract_information(self, items):
```
Your method comment still says there is a current_branch_id param.

simzacks commented on superx/information_extractor.py:
```diff
@@ -22,133 +22,139 @@ class info_extractor:
+                    continue
+
+            quantity = float(item.find('Quantity').text)
+            price = float(item.find('ItemPrice').text)
```
Please see the difference between float and decimal here:
https://docs.python.org/2/library/decimal.html

While float is faster than decimal, it is also not going to give you the exact number you want. Speed is not important in this context, as it is not the bottleneck.

simzacks commented on superx/information_extractor.py:
```diff
@@ -22,133 +22,139 @@ class info_extractor:
             update_date = self.standardize_date(item.find('PriceUpdateDate').text)
             is_weighted = False
-            if item.find(bIsWeighted).text == 1:
+            if int(item.find(bIsWeighted).text) == 1:
```
why convert to int? test for "1"

simzacks commented on superx/information_extractor.py:
```diff
@@ -22,133 +22,139 @@ class info_extractor:
+            # if item already in db then continue to next item
+            if bool(Product.query.filter_by(id=item_code).first()):
+                # adding current_branch_price to the db and commit
+                current_branch_price = BranchPrice(item_code=item_code, price=price, update_date=update_date)
```
In the code immediately after this, you add the product to the database and then the price.
Is that meant to be in an else?
In any case, If you erase this code and only add the product if it exists and then always add the price, that removes duplicate code.

Also, you probably need the BranchID in this table as well

simzacks replied:
Same question as before. Where is the branchID?

simzacks commented on superx/information_extractor.py:
```diff
@@ -22,133 +22,139 @@ class info_extractor:
+            current_product = Product(id=item_code, name=name, quantity=quantity, is_weighted=is_weighted,
+                                      unit_of_measure=unit_of_measure)
+            db.session.add(current_product)
+            db.session.commit()
```
You should probably only commit once after the next statement

simzacks commented on superx/information_extractor.py:
```diff
@@ -22,133 +22,139 @@ class info_extractor:
+            db.session.add(current_product)
+            db.session.commit()
+            # adding current_branch_price to the db and commit
+            current_branch_price = BranchPrice(item_code=item_code, price=price, update_date=update_date)
```
you probably need the BranchID in this table as well

simzacks commented on superx/models/__init__.py:
```diff
@@ -1,6 +1,11 @@
 from sqlalchemy import Integer, Column, Text, Boolean, Float
+<<<<<<< HEAD
```
This happens because of a conflict. You have to modify the file to how you want it (this removes the conflict) and then add it to the repo again before commiting

shaibrown5 replied:
not entirey sure what you mean or what to do

simzacks replied:
Modify the file so it only has the lines that you want and then commit it again

simzacks commented on superx/information_extractor.py:
```diff
@@ -161,14 +167,16 @@ def convert_unit_name(self, unit_in_hebrew):
 
-        return 'unknown'
+        # as a default return the original unit
+        return unit_in_hebrew
```
You may want to log this as well, so that you know that you need to look at it and maybe add it to the list.

simzacks reviewed (changes requested):
Looking good. definitely in the right direction

shaibrown5 committed:
Merge branch 'master' of github.com:beyond-io/superx

shaibrown5 committed:
added logging for exception handling. If no internet connection, a logger file will be made to lert us

shaibrown5 committed:
changed logging level to Info and added a logger for any new item weight.
the weight name that is logged it encoded to utf-8

simzacks commented on bootstrap.sh:
```diff
@@ -2,10 +2,12 @@
 apt-get install python3 python3-pip -y
-pip3 install flask Flask-SQLAlchemy SQLAlchemy pymysql
+pip3 install -r /vagrant/requirments.txt
+
```
Another PR removed this file. Please rebase and decide.
The pylint shows that the reason this command is failing is:
ERROR: Could not find a version that satisfies the requirement command-not-found==0.3 (from -r requirments.txt (line 17)) (from versions: none)
63
ERROR: No matching distribution found for command-not-found==0.3 (from -r requirments.txt (line 17))

simzacks commented on requirments.txt:
```diff
@@ -0,0 +1,48 @@
+chardet==2.2.1
+click==7.1.2
+colorama==0.2.5
+command-not-found==0.3
```
this is the line that pylint says it is failing on

simzacks commented on superx/information_extractor.py:
```diff
@@ -139,15 +139,20 @@ def extract_information(self, items):
+                db.session.add(current_branch_price)
+                db.session.commit()
                 continue
+
```
You only execute insert into BranchPrice if the product is already in the database. If it isn't you add the product to the database here, but don't add the Branch Price.
You probably want to check if the product exists. If not, add it. And then in any case add the branchPrice.

simzacks commented on superx/models/__init__.py:
```diff
@@ -1,14 +1,27 @@
 from sqlalchemy import Integer, Column, Text, Boolean, Float
+<<<<<<< HEAD
```
You have to remove the lines like <<<<<<< HEAD that don't belong here.

simzacks commented on superx/routes/home.py:
```diff
@@ -4,10 +4,15 @@
-    return render_template('login.html')
+    return render_template('home-signedin.jinja2', basket=basket)
 
+    
```
it shows the last line is not blank/. Probably there is a tab here.

simzacks commented on superx/routes/signup.py:
```diff
@@ -0,0 +1,49 @@
+from app import app
```
You'll want to rebase the code, this code was already changed in another PR

shaibrown5 replied:
ill look into what rebaseing means and how to do it

simzacks commented on superx/information_extractor.py:
```diff
@@ -57,7 +62,10 @@ def run_extractor(self):
-                url_list = self.get_all_super_links()
+                try:
+                    url_list = self.get_all_super_links()
+                except ConnectionError as ce:
```
is ConnectionError defined anywhere? I see it is being used a lot in this class

simzacks commented on superx/information_extractor.py:
```diff
@@ -196,6 +205,9 @@ def get_all_super_links(self):
         """
         num_of_pages = self.get_num_of_pages()
+        if num_of_pages == -1:
+            raise ConnectionError(f'Unable to connect to url to find number of pages for {self.current_super["store_name"]}')
```
where is ConnectionError defined?

shaibrown5 replied:
not sure what you mean by defined

simzacks reviewed (changes requested):
Aisde from the requested changes, this PR has conflicts that must be resolved.
Read this for details on how to do this: https://opensource.com/article/20/4/git-merge-conflict

shaibrown5 committed:
put the script in a file,
changed it so that fomr the "matrix" group only victory will be chosen

shaibrown5 committed:
changed float to decimal

shaibrown5 commented:
closed until i fix my code
